### DISCUSSION TITLE
The chat action is not opening the default browser with the chat URL on the remote system.

### DISCUSSION THREAD

**Discussion by amoljagdalepucsd** [Ylianst/MeshCentral#5645] [Q&A]
Hi Community , 
I found the one scenario where chat functionality is failing

Scenario 0:  A system with a active RDP session before initiating an Remote Desktop session.
 
- When I start the Desktop session, it displays an active session list (1. user1 2. Connect to new session). If we select 'user1' and switch to that session, attempting to start a chat by pressing the chat action on the remote system does not work. An error occurs while trying to automatically open the browser from CMD / PowerShell.
 
Scenario 1: A system with no active RDP session.
 
- In this scenario, if I attempt to initiate an Desktop session and try the chat option, it works perfectly, opening the chat browser on the remote system.
 
Therefore, when creating and running the chat task at the user session level, it fails and throws an error.

So Is it default behavior or an issue ?

**Comment by Henrique-Batista** [comment]
I read somewhere that active sessions only work in the console session, in the case when there is a user actually logged in and using the computer or when you connect via the Desktop. When you access via RDP you create a remote virtual session, and it seems that the command to open a browser or open a chat requires an active console session on the computer.

**Reply by amoljagdalepucsd** [reply]
Thank you @Henrique-Batista for response.
can you please share the link where you read that details if possible

**Comment by si458** [comment]
whats the OS of the remote device?
just tried it here with Server 2022 as VM,
connected to it via RDP from my lappy, then went into meshcentral, connected to rdp session via user select (as you said) then clicked chat icon and it opened up no problem?

**Reply by amoljagdalepucsd** [reply]
Thank you @si458 for reply.
I tried it for windows 10 laptop, Windows 11 Enterprise N as VM but its failing.
Can you please share your config.JSON (configuration file)  
I am getting this exception in openUserDesktopUrl function of meshcore.js : **ITaskService failed to register action -2147023564**

**Reply by si458** [reply]
Where are you seeing this error?
As the cmd prompt window should close instantly, so u don't see anything?

**Reply by amoljagdalepucsd** [reply]
We are using two files for Chat.
1) Meshcore.js [Server side ]
2) win-tasks.js [Agent side ]

Below function is in meshcore.js file which we are using to open browser with chat URL.
I am getting this exception in catch block (bold line) which is thrown from win-tasks.js .

```
function openUserDesktopUrl(url) {
    if ((url.toLowerCase().startsWith('http://') == false) && (url.toLowerCase().startsWith('https://') == false)) { return null; }
        var child = null;
        switch (process.platform) {
            case 'win32':
                var uid = require('user-sessions').consoleUid();
                var user = require('user-sessions').getUsername(uid);
                var domain = require('user-sessions').getDomain(uid);
                var task = { name: 'MeshChatTask', user: user, domain: domain, execPath: process.env['windir'] + '\\system32\\cmd.exe', arguments: ['/C START ' + url.split('&').join('^&')] };
                try {
                    require('win-tasks').addTask(task);
                    require('win-tasks').getTask({ name: 'MeshChatTask' }).run();
                    require('win-tasks').deleteTask('MeshChatTask');
                    return (true);
                }
                catch (ex) {
                    **sendConsoleText("CATCH ->" + ex)**
                    var taskoptions = { env: { _target: process.env['windir'] + '\\system32\\cmd.exe', _args: '/C START ' + url.split('&').join('^&'), _user: '"' + domain + '\\' + user + '"' } };
                    for (var c1e in process.env) {
                        taskoptions.env[c1e] = process.env[c1e];
                    }
                    var child = require('child_process').execFile(process.env['windir'] + '\\System32\\WindowsPowerShell\\v1.0\\powershell.exe', ['powershell', '-noprofile', '-nologo', '-command', '-'], taskoptions);
                    child.stderr.on('data', function (c) { });
                    child.stdout.on('data', function (c) { });
                    child.stdin.write('SCHTASKS /CREATE /F /TN MeshChatTask /SC ONCE /ST 00:00 ');
                    if (user) { child.stdin.write('/RU $env:_user '); }
                    child.stdin.write('/TR "$env:_target $env:_args"\r\n');
                    child.stdin.write('$ts = New-Object -ComObject Schedule.service\r\n');
                    child.stdin.write('$ts.connect()\r\n');
                    child.stdin.write('$tsfolder = $ts.getfolder("\\")\r\n');
                    child.stdin.write('$task = $tsfolder.GetTask("MeshChatTask")\r\n');
                    child.stdin.write('$taskdef = $task.Definition\r\n');
                    child.stdin.write('$taskdef.Settings.StopIfGoingOnBatteries = $false\r\n');
                    child.stdin.write('$taskdef.Settings.DisallowStartIfOnBatteries = $false\r\n');
                    child.stdin.write('$taskdef.Actions.Item(1).Path = $env:_target\r\n');
                    child.stdin.write('$taskdef.Actions.Item(1).Arguments = $env:_args\r\n');
                    child.stdin.write('$tsfolder.RegisterTaskDefinition($task.Name, $taskdef, 4, $null, $null, $null)\r\n');

                    child.stdin.write('SCHTASKS /RUN /TN MeshChatTask\r\n');
                    child.stdin.write('SCHTASKS /DELETE /F /TN MeshChatTask\r\nexit\r\n');
                    child.waitExit();
                }
                break;
            case 'linux':
                child = require('child_process').execFile('/usr/bin/xdg-open', ['xdg-open', url], 
                  { uid: require('user-sessions').consoleUid() });
                break;
            case 'darwin':
                child = require('child_process').execFile('/usr/bin/open', ['open', url], 
                  { uid: require('user-sessions').consoleUid() });
                break;
            default:
                // Unknown platform, ignore this command.
                break;
        }
}
```

**Reply by si458** [reply]
Right, so for some reason, when it tries to add the windows task to the computer to open the url, it's failing, and then falling to the backup method of using powershell to add the task, but that's ALSO failing...
So the must be some sort of restriction on the remote devices which isn't allowing mesh to add to windows task scheduler 
I think we need to add a final catch to maybe just try running cmd.exe with the open and url command maybe?

**Reply by amoljagdalepucsd** [reply]
Thanks @si458 .
But how its working for you ?
I tried final catch by running cmd.exe with **START  ""  URL** command.
I tried it in two ways :
1.  Using mesh **runcommands** functionality.
2.  By creating  simple child process to open cmd with url command 
            {
                var options = { type: 1, uid: uid };
                var child = require('child_process').execFile(process.env['windir'] + '\\system32\\cmd.exe', ['/c', **'START https://github.com/**'], options);
                child.waitExit();
            }

**Reply by si458** [reply]
oh look you have this issue - https://github.com/Ylianst/MeshCentral/issues/4884
dont reply to this going forward please but reply to the issue instead, makes things neater

**Reply by amoljagdalepucsd** [reply]
Ok Si458, I will reply to the issue #4884 .
